File: modules/data_analyzer.py
import pandas as pd
import numpy as np
from typing import Dict, Any, Optional


WEEKDAY_KO = ['월', '화', '수', '목', '금', '토', '일']


def _categorize(name: str, keyword_map: Dict[str, list]) -> str:
    name_str = str(name)
    for cat, keywords in keyword_map.items():
        for kw in keywords:
            if kw in name_str:
                return cat
    return '기타'


def _fmt_krw(v: float) -> str:
    if v >= 100_000_000:
        return f"{v / 100_000_000:.1f}억원"
    elif v >= 10_000:
        return f"{v / 10_000:.0f}만원"
    return f"{v:,.0f}원"


class DataAnalyzer:
    def __init__(self, hospital_config: dict):
        self.config = hospital_config
        self.kw_map: Dict[str, list] = hospital_config.get('category_keywords', {})

    def analyze(self, parsed_data: Dict[str, pd.DataFrame]) -> Dict[str, Any]:
        order_df: Optional[pd.DataFrame] = parsed_data.get('오더판매내역')
        cat_df:   Optional[pd.DataFrame] = parsed_data.get('분야별집계')
        ch_df:    Optional[pd.DataFrame] = parsed_data.get('내원경로별')

        results: Dict[str, Any] = {}

        if order_df is not None and not order_df.empty:
            results.update(self._analyze_order(order_df))

        if cat_df is not None and not cat_df.empty and 'category_sales' not in results:
            results['original_categories'] = self._summarize_cat(cat_df)

        if ch_df is not None and not ch_df.empty and 'channel_sales' not in results:
            results['channel_sales'] = self._summarize_channel(ch_df)

        results['key_findings'] = self._generate_findings(results)
        results['insights'] = self._generate_insights(results)

        return results

    # ── 오더판매내역 분석 ──────────────────────────────────────────────────────

    def _analyze_order(self, df: pd.DataFrame) -> Dict[str, Any]:
        r: Dict[str, Any] = {}

        # 카테고리 분류
        if 'order_name' in df.columns:
            df = df.copy()
            df['category'] = df['order_name'].apply(lambda x: _categorize(x, self.kw_map))
        elif 'original_category' in df.columns:
            df = df.copy()
            df['category'] = df['original_category']
        else:
            df = df.copy()
            df['category'] = '기타'

        # KPIs
        total_rev = df['amount'].sum() if 'amount' in df.columns else 0
        total_visits = len(df)
        patients = df['chart_no'].nunique() if 'chart_no' in df.columns else total_visits
        avg_per_patient = total_rev / patients if patients else 0
        avg_per_visit   = total_rev / total_visits if total_visits else 0

        r['kpis'] = {
            'total_revenue':   total_rev,
            'total_visits':    total_visits,
            'unique_patients': patients,
            'avg_per_patient': avg_per_patient,
            'avg_per_visit':   avg_per_visit,
        }

        # 카테고리별 매출
        cat_grp = (
            df.groupby('category', as_index=False)['amount']
              .agg(revenue='sum', count='count')
              .sort_values('revenue', ascending=False)
        )
        cat_grp['pct'] = cat_grp['revenue'] / cat_grp['revenue'].sum() * 100
        r['category_sales'] = cat_grp

        # 기존(CRM) vs 재분류 비교
        if 'original_category' in df.columns:
            orig = (
                df.groupby('original_category', as_index=False)['amount']
                  .agg(orig_revenue='sum', orig_count='count')
            )
            new_ = (
                df.groupby('category', as_index=False)['amount']
                  .agg(new_revenue='sum', new_count='count')
                  .rename(columns={'category': 'original_category'})
            )
            r['reclassify_compare'] = orig.merge(new_, on='original_category', how='outer').fillna(0)

        # 진료의별
        if 'doctor' in df.columns:
            doc_grp = (
                df.groupby(['doctor', 'category'], as_index=False)['amount']
                  .agg(revenue='sum', count='count')
                  .sort_values('revenue', ascending=False)
            )
            r['doctor_sales'] = doc_grp

        # 유입경로별
        if 'channel' in df.columns:
            ch_grp = (
                df.groupby('channel', as_index=False)['amount']
                  .agg(revenue='sum', count='count')
                  .sort_values('revenue', ascending=False)
            )
            ch_grp['avg'] = ch_grp['revenue'] / ch_grp['count']
            r['channel_sales'] = ch_grp

            # 채널 × 카테고리
            try:
                pivot = df.groupby(['channel', 'category'])['amount'].sum().unstack(fill_value=0)
                r['channel_x_category'] = pivot
            except Exception:
                pass

        # 국적별
        if 'nationality' in df.columns:
            nat_grp = (
                df.groupby('nationality', as_index=False)['amount']
                  .agg(revenue='sum', count='count')
                  .sort_values('revenue', ascending=False)
            )
            nat_grp['avg'] = nat_grp['revenue'] / nat_grp['count']
            r['nationality_sales'] = nat_grp

        # 환자 객단가
        if 'chart_no' in df.columns:
            per_pt = (
                df.groupby('chart_no')['amount']
                  .agg(total='sum', visits='count')
            )
            r['patient_spending'] = per_pt

        # 복합시술 (Cross-selling)
        if 'chart_no' in df.columns:
            pt_counts = df.groupby('chart_no')['category'].nunique()
            cross_rate = (pt_counts >= 2).mean() * 100
            single_avg = per_pt.loc[pt_counts == 1, 'total'].mean() if (pt_counts == 1).any() else 1
            multi_avg  = per_pt.loc[pt_counts >= 2, 'total'].mean() if (pt_counts >= 2).any() else 0
            multiplier = multi_avg / single_avg if single_avg else 0

            # TOP 조합
            combos = (
                df.groupby('chart_no')['category']
                  .apply(lambda x: tuple(sorted(set(x))))
                  .reset_index(name='combo')
            )
            top_combos = (
                combos[combos['combo'].apply(len) >= 2]
                ['combo']
                .value_counts()
                .head(5)
                .reset_index()
            )
            top_combos.columns = ['조합', '횟수']
            top_combos['조합'] = top_combos['조합'].apply(lambda x: ' + '.join(x))

            r['cross_selling'] = {
                'cross_sell_rate': cross_rate,
                'avg_multiplier':  multiplier,
                'top_combinations': top_combos,
            }

        # 일별 / 요일별 추이
        if 'date' in df.columns:
            df_dated = df.dropna(subset=['date'])
            if not df_dated.empty:
                daily = (
                    df_dated.groupby(df_dated['date'].dt.date)['amount']
                            .agg(revenue='sum', count='count')
                            .reset_index()
                )
                daily.columns = ['date', 'revenue', 'count']
                r['daily_trends'] = daily

                df_dated = df_dated.copy()
                df_dated['weekday'] = df_dated['date'].dt.dayofweek
                wd = (
                    df_dated.groupby('weekday')['amount']
                            .agg(revenue='sum', count='count')
                            .reindex(range(7), fill_value=0)
                            .reset_index()
                )
                wd['avg'] = wd['revenue'] / wd['count'].replace(0, np.nan)
                wd['weekday_name'] = wd['weekday'].map(dict(enumerate(WEEKDAY_KO)))
                r['weekday_trends'] = wd

        # 리프팅 기기별
        if 'order_name' in df.columns:
            lifting_kws = self.kw_map.get('리프팅', [])
            lift_df = df[df['category'] == '리프팅'].copy() if '리프팅' in df['category'].values else pd.DataFrame()
            if not lift_df.empty and 'order_name' in lift_df.columns:
                lift_grp = (
                    lift_df.groupby('order_name')['amount']
                           .agg(revenue='sum', count='count')
                           .sort_values('revenue', ascending=False)
                           .head(15)
                           .reset_index()
                )
                r['lifting_detail'] = lift_grp

            # 기타 세부
            misc_df = df[df['category'] == '기타'].copy() if '기타' in df['category'].values else pd.DataFrame()
            if not misc_df.empty and 'order_name' in misc_df.columns:
                misc_grp = (
                    misc_df.groupby('order_name')['amount']
                           .agg(revenue='sum', count='count')
                           .sort_values('revenue', ascending=False)
                           .head(20)
                           .reset_index()
                )
                r['misc_detail'] = misc_grp

        return r

    # ── 분야별집계 파싱 ────────────────────────────────────────────────────────

    def _summarize_cat(self, df: pd.DataFrame) -> pd.DataFrame:
        if 'original_category' not in df.columns or 'amount' not in df.columns:
            return df
        grp = (
            df.groupby('original_category', as_index=False)['amount']
              .sum()
              .sort_values('amount', ascending=False)
        )
        grp['pct'] = grp['amount'] / grp['amount'].sum() * 100
        return grp

    def _summarize_channel(self, df: pd.DataFrame) -> pd.DataFrame:
        ch_col = 'channel' if 'channel' in df.columns else 'original_category'
        if ch_col not in df.columns or 'amount' not in df.columns:
            return df
        grp = (
            df.groupby(ch_col, as_index=False)['amount']
              .agg(revenue='sum', count='count' if 'count' in df.columns else lambda x: len(x))
              .sort_values('revenue', ascending=False)
        )
        return grp

    # ── 인사이트 생성 ─────────────────────────────────────────────────────────

    def _generate_findings(self, r: Dict) -> list:
        findings = []
        try:
            if 'category_sales' in r and not r['category_sales'].empty:
                top = r['category_sales'].iloc[0]
                findings.append(f"가장 높은 매출 카테고리: {top['category']} ({top['pct']:.1f}%)")

            if 'channel_sales' in r and not r['channel_sales'].empty:
                top_ch = r['channel_sales'].iloc[0]
                findings.append(f"주요 유입경로: {top_ch['channel']} (방문 {top_ch['count']:,}건)")

            if 'cross_selling' in r:
                cs = r['cross_selling']
                findings.append(f"복합시술 비율: {cs['cross_sell_rate']:.1f}% (단일 대비 {cs['avg_multiplier']:.1f}배 객단가)")

            if 'kpis' in r:
                kpis = r['kpis']
                findings.append(f"평균 객단가: {_fmt_krw(kpis['avg_per_patient'])} (총 {kpis['unique_patients']:,}명)")
        except Exception:
            pass
        return findings

    def _generate_insights(self, r: Dict) -> Dict[str, list]:
        insights = {'원장님 관점': [], '마케팅 관점': [], '운영 관점': []}
        try:
            if 'category_sales' in r and not r['category_sales'].empty:
                top2 = r['category_sales'].head(2)
                cats = ' / '.join(top2['category'].tolist())
                insights['원장님 관점'].append(f"핵심 매출 시술: {cats} — 집중 역량 강화 권고")
                insights['원장님 관점'].append("고객당 평균 내원 횟수 기준으로 재방문 시술 패키지 기획 검토")

            if 'channel_sales' in r and not r['channel_sales'].empty:
                top_ch = r['channel_sales'].iloc[0]
                insights['마케팅 관점'].append(f"주요 채널 '{top_ch['channel']}' 집중 투자로 ROI 극대화")
                if len(r['channel_sales']) > 1:
                    ch2 = r['channel_sales'].iloc[1]
                    insights['마케팅 관점'].append(f"2위 채널 '{ch2['channel']}' 성장 가능성 테스트 권장")

            if 'patient_spending' in r:
                pt = r['patient_spending']
                p90 = pt['total'].quantile(0.9)
                insights['원장님 관점'].append(f"상위 10% 고객 객단가: {_fmt_krw(p90)} — VIP 케어 프로그램 도입 검토")

            if 'weekday_trends' in r and not r['weekday_trends'].empty:
                wt = r['weekday_trends']
                best_day = wt.loc[wt['revenue'].idxmax(), 'weekday_name']
                worst_day = wt.loc[wt['revenue'].idxmin(), 'weekday_name']
                insights['운영 관점'].append(f"최다 매출 요일: {best_day}요일 — 스태프 배치 강화 권고")
                insights['운영 관점'].append(f"최저 매출 요일: {worst_day}요일 — 프로모션/패키지 행사 기획")

            insights['운영 관점'].append("내원 경로별 예약 시스템 최적화로 대기 시간 단축 검토")
        except Exception:
            pass
        return insights

File: frontend/src/lib/dataAnalyzer.ts
import { OrderRow, ParsedData } from './excelParser';
import { HospitalConfig, categorize } from './hospitalConfig';

export interface KPIs {
  totalRevenue: number;
  totalVisits: number;
  uniquePatients: number;
  avgPerPatient: number;
  avgPerVisit: number;
}

export interface CategorySales {
  category: string; revenue: number; count: number; pct: number;
}
export interface ProcedureSales {
  name: string; revenue: number; count: number; pct: number; category: string;
}
export interface DoctorSales {
  doctor: string; category: string; revenue: number; count: number;
}
export interface ChannelSales {
  channel: string; revenue: number; count: number; avg: number;
}
export interface NationalitySales {
  nationality: string; revenue: number; count: number; avg: number;
}
export interface CrossSelling {
  crossSellRate: number;
  avgMultiplier: number;
  topCombinations: { combo: string; count: number }[];
}
export interface DailyTrend { date: string; revenue: number; count: number; }
export interface WeekdayTrend { name: string; revenue: number; count: number; avg: number; }
export interface PatientSpend { chartNo: string; total: number; visits: number; }

export interface AnalysisResult {
  kpis: KPIs;
  categorySales: CategorySales[];
  topProcedures: ProcedureSales[];
  doctorSales: DoctorSales[];
  channelSales: ChannelSales[];
  nationalitySales: NationalitySales[];
  crossSelling: CrossSelling | null;
  dailyTrends: DailyTrend[];
  weekdayTrends: WeekdayTrend[];
  patientSpending: PatientSpend[];
  keyFindings: string[];
  insights: { doctor: string[]; marketing: string[]; operations: string[] };
}

const WEEKDAY = ['월', '화', '수', '목', '금', '토', '일'];

function groupBy<T>(arr: T[], key: (x: T) => string): Map<string, T[]> {
  const m = new Map<string, T[]>();
  for (const x of arr) {
    const k = key(x);
    if (!m.has(k)) m.set(k, []);
    m.get(k)!.push(x);
  }
  return m;
}

function sumBy<T>(arr: T[], fn: (x: T) => number) {
  return arr.reduce((s, x) => s + fn(x), 0);
}

function fmtKrw(v: number): string {
  if (v >= 1e8) return `${(v / 1e8).toFixed(1)}억원`;
  if (v >= 1e4) return `${Math.round(v / 1e4)}만원`;
  return `${v.toLocaleString()}원`;
}

export class DataAnalyzer {
  constructor(private cfg: HospitalConfig) {}

  analyze(data: ParsedData): AnalysisResult {
    const orders = data.orders;
    const result = {} as AnalysisResult;

    // Categorize
    const tagged = orders.map(r => ({
      ...r,
      category: categorize(r.order_name, this.cfg.categoryKeywords),
    }));

    // KPIs
    if (tagged.length) {
      const totalRevenue   = sumBy(tagged, r => r.amount);
      const totalVisits    = tagged.length;
      const patients       = new Set(tagged.filter(r => r.chart_no).map(r => r.chart_no)).size || totalVisits;
      result.kpis = {
        totalRevenue,
        totalVisits,
        uniquePatients: patients,
        avgPerPatient:  totalRevenue / patients,
        avgPerVisit:    totalRevenue / totalVisits,
      };
    } else {
      result.kpis = { totalRevenue: 0, totalVisits: 0, uniquePatients: 0, avgPerPatient: 0, avgPerVisit: 0 };
    }

    // Category sales
    const catGroups = groupBy(tagged, r => r.category);
    const totalRev = result.kpis.totalRevenue || 1;
    result.categorySales = Array.from(catGroups.entries())
      .map(([category, rows]) => ({
        category, revenue: sumBy(rows, r => r.amount), count: rows.length,
        pct: (sumBy(rows, r => r.amount) / totalRev) * 100,
      }))
      .sort((a, b) => b.revenue - a.revenue);

    // Top individual procedures (시술별 TOP 순위)
    const procGroups = groupBy(tagged.filter(r => r.order_name), r => r.order_name);
    result.topProcedures = Array.from(procGroups.entries())
      .map(([name, rows]) => {
        const revenue = sumBy(rows, r => r.amount);
        return {
          name,
          revenue,
          count: rows.length,
          pct: (revenue / totalRev) * 100,
          category: rows[0].category,
        };
      })
      .sort((a, b) => b.revenue - a.revenue)
      .slice(0, 20);

    // Doctor sales
    const docRows = tagged.filter(r => r.doctor && r.doctor !== '미기재');
    if (docRows.length) {
      const pairs = groupBy(docRows, r => `${r.doctor}___${r.category}`);
      result.doctorSales = Array.from(pairs.entries())
        .map(([key, rows]) => {
          const [doctor, category] = key.split('___');
          return { doctor, category, revenue: sumBy(rows, r => r.amount), count: rows.length };
        })
        .sort((a, b) => b.revenue - a.revenue);
    } else {
      result.doctorSales = [];
    }

    // Channel sales
    const chRows = tagged.filter(r => r.channel && r.channel !== '미기재');
    if (chRows.length) {
      const chGroups = groupBy(chRows, r => r.channel);
      result.channelSales = Array.from(chGroups.entries())
        .map(([channel, rows]) => {
          const revenue = sumBy(rows, r => r.amount);
          return { channel, revenue, count: rows.length, avg: revenue / rows.length };
        })
        .sort((a, b) => b.revenue - a.revenue);
    } else if (data.channelSummary.length) {
      result.channelSales = data.channelSummary.map(r => ({
        channel: r.channel, revenue: r.amount, count: 0, avg: 0,
      })).sort((a, b) => b.revenue - a.revenue);
    } else {
      result.channelSales = [];
    }

    // Nationality
    const natRows = tagged.filter(r => r.nationality && r.nationality !== '미기재');
    if (natRows.length) {
      const natGroups = groupBy(natRows, r => r.nationality);
      result.nationalitySales = Array.from(natGroups.entries())
        .map(([nationality, rows]) => {
          const revenue = sumBy(rows, r => r.amount);
          return { nationality, revenue, count: rows.length, avg: revenue / rows.length };
        })
        .sort((a, b) => b.revenue - a.revenue);
    } else {
      result.nationalitySales = [];
    }

    // Patient spending
    const ptGroups = groupBy(tagged.filter(r => r.chart_no), r => r.chart_no);
    result.patientSpending = Array.from(ptGroups.entries())
      .map(([chartNo, rows]) => ({
        chartNo, total: sumBy(rows, r => r.amount), visits: rows.length,
      }))
      .sort((a, b) => b.total - a.total);

    // Cross-selling
    if (ptGroups.size > 0) {
      const ptCatCounts = Array.from(ptGroups.entries()).map(([, rows]) => {
        const cats = new Set(rows.map(r => r.category));
        return { total: sumBy(rows, r => r.amount), catCount: cats.size, cats: [...cats].sort() };
      });
      const multi = ptCatCounts.filter(p => p.catCount >= 2);
      const single = ptCatCounts.filter(p => p.catCount === 1);
      const crossRate = ptGroups.size > 0 ? (multi.length / ptGroups.size) * 100 : 0;
      const singleAvg = single.length ? sumBy(single, p => p.total) / single.length : 1;
      const multiAvg  = multi.length  ? sumBy(multi,  p => p.total) / multi.length  : 0;

      const comboMap = new Map<string, number>();
      for (const p of multi) {
        const key = p.cats.join(' + ');
        comboMap.set(key, (comboMap.get(key) || 0) + 1);
      }
      const topCombos = [...comboMap.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
        .map(([combo, count]) => ({ combo, count }));

      result.crossSelling = {
        crossSellRate: crossRate,
        avgMultiplier: singleAvg > 0 ? multiAvg / singleAvg : 0,
        topCombinations: topCombos,
      };
    } else {
      result.crossSelling = null;
    }

    // Daily trends
    const datedRows = tagged.filter(r => r.date instanceof Date);
    if (datedRows.length) {
      const dayGroups = groupBy(datedRows, r => (r.date as Date).toISOString().slice(0, 10));
      result.dailyTrends = [...dayGroups.entries()]
        .map(([date, rows]) => ({ date, revenue: sumBy(rows, r => r.amount), count: rows.length }))
        .sort((a, b) => a.date.localeCompare(b.date));

      const wdGroups = groupBy(datedRows, r => String((r.date as Date).getDay()));
      result.weekdayTrends = Array.from({ length: 7 }, (_, i) => {
        const day = (i + 1) % 7;
        const rows = wdGroups.get(String(day)) || [];
        const revenue = sumBy(rows, r => r.amount);
        return { name: WEEKDAY[i], revenue, count: rows.length, avg: rows.length ? revenue / rows.length : 0 };
      });
    } else {
      result.dailyTrends = [];
      result.weekdayTrends = [];
    }

    result.keyFindings = this.generateFindings(result);
    result.insights     = this.generateInsights(result);
    return result;
  }

  private generateFindings(r: AnalysisResult): string[] {
    const findings: string[] = [];
    if (r.categorySales.length) {
      const top = r.categorySales[0];
      findings.push(`최고 매출 카테고리: ${top.category} (${top.pct.toFixed(1)}% · ${fmtKrw(top.revenue)})`);
    }
    if (r.topProcedures.length) {
      const top = r.topProcedures[0];
      findings.push(`매출 1위 시술: ${top.name} (${fmtKrw(top.revenue)} · ${top.count}건)`);
    }
    if (r.channelSales.length) {
      const top = r.channelSales[0];
      findings.push(`주요 유입경로: ${top.channel} (방문 ${top.count.toLocaleString()}건)`);
    }
    if (r.crossSelling) {
      const cs = r.crossSelling;
      findings.push(`복합시술 비율: ${cs.crossSellRate.toFixed(1)}% · 단일 대비 ${cs.avgMultiplier.toFixed(1)}배 객단가`);
    }
    findings.push(`평균 객단가: ${fmtKrw(r.kpis.avgPerPatient)} (총 ${r.kpis.uniquePatients.toLocaleString()}명)`);
    return findings;
  }

  private generateInsights(r: AnalysisResult): AnalysisResult['insights'] {
    const ins = { doctor: [] as string[], marketing: [] as string[], operations: [] as string[] };

    if (r.categorySales.length >= 2) {
      const top2 = r.categorySales.slice(0, 2).map(c => c.category).join(' / ');
      ins.doctor.push(`핵심 매출 시술 ${top2} — 집중 역량 강화 권고`);
    }
    if (r.topProcedures.length) {
      ins.doctor.push(`매출 1위 시술 '${r.topProcedures[0].name}' 수요 강세 지속 — 예약 슬롯 확대 검토`);
    }
    if (r.patientSpending.length >= 10) {
      const p90 = r.patientSpending[Math.floor(r.patientSpending.length * 0.1)].total;
      ins.doctor.push(`상위 10% 객단가 ${fmtKrw(p90)} — VIP 관리 프로그램 도입 검토`);
    }

    if (r.channelSales.length >= 1) {
      const topVol = r.channelSales[0];
      ins.marketing.push(`유입량 1위 '${topVol.channel}' (${topVol.count}건) — 예산 유지 및 광고 소재 리프레시 권장`);
    }
    if (r.channelSales.length >= 2) {
      const byEff = [...r.channelSales].filter(c => c.count >= 3).sort((a, b) => b.avg - a.avg);
      if (byEff.length) {
        ins.marketing.push(`건당 매출 최고 채널 '${byEff[0].channel}' (${fmtKrw(byEff[0].avg)}) — 투자 확대 우선 고려`);
      }
    }
    if (r.crossSelling && r.crossSelling.crossSellRate < 30) {
      ins.marketing.push(`복합시술 비율 ${r.crossSelling.crossSellRate.toFixed(1)}% — 패키지 프로모션으로 객단가 제고 가능`);
    }

    if (r.weekdayTrends.length) {
      const best  = r.weekdayTrends.reduce((a, b) => a.revenue > b.revenue ? a : b);
      const worst = r.weekdayTrends.filter(d => d.count > 0).reduce((a, b) => a.revenue < b.revenue ? a : b);
      ins.operations.push(`최다 매출 ${best.name}요일 — 스태프 배치 강화 및 예약 시스템 최적화`);
      if (worst.name !== best.name) {
        ins.operations.push(`${worst.name}요일 매출 저조 — 한정 이벤트 기획으로 방문 유도`);
      }
    }
    if (r.kpis.uniquePatients > 0) {
      const revisitRate = r.patientSpending.filter(p => p.visits >= 2).length / r.kpis.uniquePatients * 100;
      if (revisitRate > 0) {
        ins.operations.push(`재방문 환자 비율 ${revisitRate.toFixed(1)}% — 리텐션 프로그램 강화 권장`);
      }
    }
    return ins;
  }
}
